core/primitives/src/lib.rs

```diff
@@ -80,6 +80,35 @@ pub use self::hasher::blake2::Blake2Hasher;
 /// A 512-bit value interpreted as a signature.
 pub type Signature = hash::H512;
 
+/// Context for executing a call into the runtime.
+#[repr(u8)]
+pub enum ExecutionContext {
+	/// Context for general importing (including own blocks).
+	Importing,
+	/// Context used when syncing the blockchain.
+	Syncing,
+	/// Context used for block construction.
+	BlockConstruction,
+	/// Offchain worker context.
+	OffchainWorker(Box<OffchainExt>),
+	/// Context used for other calls.
+	Other,
+}
+
+/// An extended externalities for offchain workers.
+pub trait OffchainExt {
+	/// Submits an extrinsics.
+	///
+	/// The extrinsic will either go to the pool (signed)
+	/// or to the next produced block (inherent).
+	fn submit_extrinsic(&mut self, extrinsic: Vec<u8>);
+}
+impl<T: OffchainExt + ?Sized> OffchainExt for Box<T> {
```

Comment:
Yeah, but `DerefMut` just means that we can use functions from that trait directly on `Box<T>`, we need to pass `Box<T>` to a generic function accepting `T: OffchainExt` though.

I wasn't sure about the API initially and chose to handle generic parameters instead of `Box`, but currently the only way we actually pass the offchain extensions is by `ExecutionContext`, and making it generic was introducing too much noise in the code so I chose to `Box`, but only there.

Let me know if you think it's better to change the other apis and structs to store `Box` instead of generic param.